src/AstGen.zig

```diff
@@ -1451,10 +1455,13 @@ pub fn comptimeExpr(
     rl: ResultLoc,
     node: ast.Node.Index,
 ) InnerError!Zir.Inst.Ref {
-    const prev_force_comptime = gz.force_comptime;
+    const astgen = gz.astgen;
+    if (gz.force_comptime) {
+        return astgen.failNode(node, "comptime here happens inside a comptime scope; it is redundant", .{});
+    }
     gz.force_comptime = true;
     const result = try expr(gz, scope, rl, node);
-    gz.force_comptime = prev_force_comptime;
+    gz.force_comptime = false;
```

Comment:
we need a `comptimeExprFromAst` for the one that comes from the `comptime` keyword, and a `comptimeExpr` for other callsites. Otherwise, we'll get false positives for this for this code for example: `comptime { _ = @field(a, "foo"); }`. I think if you look at all the callsites of `comptimeExpr` you will see what I mean.

Be sure to note this in the doc comments so that future contributors don't accidentally call the wrong one :)